concrete/single_pages/dashboard/system/api/settings.php

```diff
@@ -41,4 +35,32 @@
         <hr>
         <div><a href="<?=URL::to('/dashboard/system/api/integrations', 'add')?>" class="btn btn-default btn-xs"><?=t('Add Integration')?></a></div>
     </section>
+
+    <section>
+        <h3><?=t('Enabled Grant Types')?></h3>
+        <div class="help-block"><?=t('Choose which OAuth2 grant types you want to have access to the API for this site.')?></div>
+
+        <?php foreach($availableGrantTypes as $key => $label) {
+            $enabled = false;
+            if (isset($grantTypes[$key]) && $grantTypes[$key] === true) {
+                $enabled = true;
+            }
+            ?>
+            <div class="checkbox">
+                <label>
+                    <?=$form->checkbox('enabledGrantTypes[]', $key, $enabled)?>
+                    <?=$label?>
```

Comment:
@aembler @mlocati I think we should apply the 'never trust user input' also to the translation system. There's currently no way to assure that translations don't contain malicious input.

Here's an example:

![xss](https://user-images.githubusercontent.com/1431100/53355150-af0f4780-3928-11e9-8565-7bce781c1b99.png)

As there are [quite some](https://translate.concrete5.org/translate/package/concrete5) languages, it's quite easy to join a team and approve strings.

I think that in case we _can_ prevent this (plain text), we should.

Reply:
PS. Dutch users don't have to be scared, this was just to illustrate the example ;)

Reply:
@a3020
1. if you are referring to strings inserted in the "Translate Site Interface" dashboard page: if an account has access to that page, it's likely that they can also edit the site (so they can insert any xss stuff they want)
2. if you are referring to strings added via translate.concrete5.org: that's why we have team coordinators: they are trusted people that have to approve strings added by normal translators...

Reply:
I think you miss my point @mlocati. I am a coordinator so I can approve strings. If my account is hacked (quite reasonable because C5 doesn't have 2FA), thousands of websites are at risk. By 'not trusting user input' you mitigate the risk as much as possible.

Reply:
Couldn't the translate.concrete5.org server sanitize the strings on download much better than running every t() function through h() on the client side?

Reply:
@mlocati to further explain the issue:

![image](https://user-images.githubusercontent.com/1431100/53357776-ebde3d00-392e-11e9-87f5-382a97b9eb35.png)

(note that this translation is NOT approved)

![image](https://user-images.githubusercontent.com/1431100/53357790-f3054b00-392e-11e9-9079-ecb75d6c0d01.png)

(note that this translation is actually used after I installed the Dutch language)

Reply:
I'd expect that team coordinators approve strings submitted by translators, or that harmful strings are replaced with good ones.
If a string is apprroved, simple translators can't replace it, they can just suggest to replace it.

In any case, if we are going to do it, I think that sanitization should be a bit smart.

For example, if an English string contains a "<script", the translation should also contain "<script".
Using "<script" in translations may be unlikely, but we can't limit to "<script": also "<svg" could be used for harmful purposes, as well as stuff like "<img onerror=".
Furthermore, a string may be contain a code snippet that's sent to the browser enlosed in `h()`.
For example, in the following case:

<?= h(t('You can't use the <script> tags')) ?>

it would be totally legit to have "<script>" in the translation.

So, I guess that the safest solution could the following one:

- when submitting a translation for a string, we detect the tags included in the English source string
- we then detect the tags included in the translated string
- the translated string can't contain tags that are not included in the source string

Reply:
@a3020 I just updated translate.concrete5.org with https://github.com/concrete5/addon_community_translation/commit/cf0e80176d823a9f58b2004defcaae7289b7afdd: now translators shoudn't be able to add potentially harmful translations.

Reply:
@mlocati I checked it, but that method can easily be circumvented, I think (as often the case with blacklists).

![image](https://user-images.githubusercontent.com/1431100/53410283-a029a300-39c3-11e9-855a-d4e4cfc96b12.png)

Reply:
@a3020 Any advice?

Reply:
@a3020 what about https://github.com/concrete5/addon_community_translation/commit/26e382024019aed4948ed9a2fb0ae1c87c663f0f ?

Reply:
@mlocati I'm not sure if that's a good solution that character list is not XSS proof. Also, are we never going to use translations with e.g. '<', '=', or '"' in it? For example, we have a few strings with `<a href="some-link"` in it, right?

@aembler Is there a (performance?) reason why you don't want to use `h` for all plain text occurrences? Escaping is the safest solution, imo.

@mlocati I found this on the Drupal site:

> Exported versions of approved translations are regularly saved, so **once your suggestions are approved**, they will be available in the exported translation files shortly.

Some more searching and I stumbled on this thread: https://www.drupal.org/project/drupal/issues/2512466 It seems they had this problem also, but then 4 yrs ago ;) I think the bottom line is that it's hard to prevent XSS.

At the moment, unapproved strings are used in production if there's no approved alternative available. I think that can be improved. Some (brainstorm!) ideas:

- **Only pull approved strings.** (this is what Drupal does, and I think it makes sense)
- Wait x-hours / x-days before an unapproved string can be pulled from the translation system.
- Having a strong policy of who can join a team / submit translations. (karma minimum?)

Reply:
@a3020 Let's continue the discussion at https://github.com/concrete5/addon_community_translation/issues/33 (a more appropriate place for CT stuff)